### ISSUE TITLE
OpenSSL error on make sdk

### ISSUE THREAD

**Issue by marioolf** [intel/linux-sgx#981]
So I am trying to update my server to version 2.21 from 2.19. Following the installation process, once I reach the `make sdk` step, I always get this error.

## Steps made
```
git clone https://github.com/intel/linux-sgx.git
cd linux-sgx && make preparation
sudo cp external/toolset/ubuntu20.04/* /usr/local/bin
make sdk
```
## Error
```
In file included from /usr/include/openssl/aes.h:14,
                 from crypto/pcl_crypto.cpp:32:
/usr/include/openssl/macros.h:138:26: warning: "OPENSSL_CONFIGURED_API" is not defined, evaluates to 0 [-Wundef]
  138 | # if OPENSSL_API_LEVEL > OPENSSL_CONFIGURED_API
      |                          ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/openssl/macros.h:155:4: error: #error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
  155 | #  error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
      |    ^~~~~
make[4]: *** [Makefile:147: crypto/pcl_crypto.o] Error 1
make[4]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/sdk/protected_code_loader'
make[3]: *** [Makefile:107: all] Error 2
make[3]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/sdk/protected_code_loader'
make[2]: *** [Makefile.source:222: sgx_pcl] Error 2
make[2]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/sdk'
make[1]: *** [Makefile:50: all] Error 2
make[1]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/sdk'
make: *** [Makefile:72: sdk] Error 2
```

## Environment
I updated mi OpenSSL version to OpenSSL 1.1.1u but nothing changed. I am using Ubuntu 22.04 also.

**Comment by llly** [comment]
The default OpenSSL in Ubuntu 2204 should be OpenSSL3 instead of OpenSSL1.1.
You can `cd sdk/protected_code_loader` then `CXXFLAGS+=-MMD make;cat crypto/pcl_crypto.d` to check your includes.
In my Ubuntu 2204, the output contains
```
crypto/pcl_crypto.o: crypto/pcl_crypto.cpp /usr/include/openssl/aes.h \
 /usr/include/openssl/macros.h \
 /usr/include/x86_64-linux-gnu/openssl/opensslconf.h \
 /usr/include/x86_64-linux-gnu/openssl/configuration.h \
 /usr/include/openssl/opensslv.h \
```
and `OPENSSL_CONFIGURED_API` is defined in `/usr/include/x86_64-linux-gnu/openssl/configuration.h`.

If your includes are different, you can try to install OpenSSL3 by `sudo apt install libssl-dev`.

**Comment by marioolf** [comment]
Solved. Compiler couldn't reach configuration.h, I had to create a sym link from  /usr/include/x86_64-linux-gnu/openssl/configuration.h to /usr/include/openssl and the problem was solved.

Thanks.

**Comment by marioolf** [comment]
New error probably related with the last, when I try to build Intel PSW with `make psw`:

```
[ 13%] Building CXX object utils/CMakeFiles/utils.dir/ssl_compat_wrapper.cpp.o
[ 14%] Building CXX object utils/CMakeFiles/utils.dir/type_length_value.cpp.o
[ 14%] Building CXX object utils/CMakeFiles/utils.dir/home/admin-ubuntu/sdks/linux-sgx/common/src/sgx_read_rand.cpp.o
[ 15%] Building CXX object utils/CMakeFiles/utils.dir/home/admin-ubuntu/sdks/linux-sgx/common/src/sgx_memset_s.cpp.o
[ 16%] Linking CXX shared library ../bin/libutils.so
/usr/local/bin/ld: CMakeFiles/utils.dir/crypto_rsa.cpp.o: in function `rsa3072_verify':
crypto_rsa.cpp:(.text+0x1ff): undefined reference to `EVP_DigestVerifyUpdate'
/usr/local/bin/ld: CMakeFiles/utils.dir/crypto_rsa.cpp.o: in function `sgx_rsa_pub_encrypt_sha256':
crypto_rsa.cpp:(.text+0x301): undefined reference to `EVP_PKEY_CTX_set_rsa_padding'
/usr/local/bin/ld: crypto_rsa.cpp:(.text+0x311): undefined reference to `EVP_PKEY_CTX_set_rsa_oaep_md'
collect2: error: ld returned 1 exit status
make[5]: *** [utils/CMakeFiles/utils.dir/build.make:277: bin/libutils.so] Error 1
make[5]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/psw/ae/aesm_service/source/build'
make[4]: *** [CMakeFiles/Makefile2:339: utils/CMakeFiles/utils.dir/all] Error 2
make[4]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/psw/ae/aesm_service/source/build'
make[3]: *** [Makefile:91: all] Error 2
make[3]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/psw/ae/aesm_service/source/build'
make[2]: *** [Makefile:79: source/build/bin/aesm_service] Error 2
make[2]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/psw/ae/aesm_service'
make[1]: *** [Makefile:61: AESM] Error 2
make[1]: Leaving directory '/home/admin-ubuntu/sdks/linux-sgx/psw/ae'
make: *** [Makefile:52: AE] Error 2
```

Openssl version is 3.0.2 and libraries where installed again with `sudo apt install libssl-dev`

Output of `cd sdk/protected_code_loader`  `CXXFLAGS+=-MMD make;cat crypto/pcl_crypto.d` is:

```
crypto/pcl_crypto.o: crypto/pcl_crypto.cpp /usr/include/openssl/aes.h \
 /usr/include/openssl/macros.h /usr/include/openssl/opensslconf.h \
 /usr/include/openssl/configuration.h /usr/include/openssl/opensslv.h \
```

**Comment by llly** [comment]
Maybe you have wrong customized including path. `/usr/include/x86_64-linux-gnu/` should be in including path.
```
$ g++ -v -x c -E -
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/11/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

Your send failure is also caused by including path. `aesm_service` from PSW uses prebuilt openssl1.1.1 and `EVP_DigestVerifyUpdate` locates in `external/dcap_source/prebuilt/openssl/inc/openssl/evp.h`.

**Comment by marioolf** [comment]
`g++ -v -x c -E -` command apparently shows correct info...

```
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/11/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

**Comment by marioolf** [comment]
Solved with 2.22 update
